File: 2 Metaheurísticas con base en una solución/Códigos/Random_Search.m
clear all
close all
clc

f = @(x,y) (x-2).^2+(y-2).^2;
xl = [-5 -5]';
xu = [5 5]';

% f = @(x,y) x.*exp(-x.^2-y.^2);
% xl = [-2 -2]';
% xu = [2 2]';

G = 500;
D = 2;

x = [4 4]';

f_plot = zeros(1,G);

for g=1:G
    y = xl + (xu-xl).*rand(D,1);

    if f(y(1),y(2)) < f(x(1),x(2))
        x = y;
    end
    
%     Plot_Contour(f,[x y],xl,xu) % Gráfica
    f_plot(g) = f(x(1),x(2));
end

figure
Plot_Surf(f,x,xl,xu) % Gráfica
disp([' mínimo global en: x=' num2str(x(1)) ', y=' num2str(x(2)) ', f(x,y)=' num2str(f(x(1),x(2)))])

figure
hold on
grid on
plot(f_plot,'b-','LineWidth',2)
title('Gráfica de convergencia')
xlabel('iteración')
ylabel('f(x)')

File: 2 Metaheurísticas con base en una solución/Códigos/Random_Search.py
import matplotlib.pyplot as plt
import numpy as np
from Plot_Surf import plot_surf

f = lambda x, y: (x-2)**2 + (y-2)**2

xl = np.array([-5, -5])
xu = np.array([5, 5])
x = np.array([4, 4])

G = 500
f_plot = np.zeros(G)

for i in range(G):
    y = xl + (xu-xl) * np.random.random(2)

    if f(y[0], y[1]) < f(x[0], x[1]):
        x = y

    f_plot[i] = f(x[0], x[1])

plot_surf(f,x,xl,xu)
print("Mínimo global en x=", x[0], " y=", x[1], " f(x,y)=", f(x[0],x[1]))

plt.plot(range(G), f_plot)
plt.title("Convergencia")
plt.show()
